### ISSUE TITLE
The transcription speed is much slower than the official demo of Whisper

### ISSUE THREAD

**Issue by LinWhite2333** [krikristoophe/whisper_flutter_plus#10]
<img src="https://github.com/krikristoophe/whisper_flutter_plus/assets/59436976/9fa92100-dd28-43de-8b64-8644dc5c1432" width=400>
<img src="https://github.com/krikristoophe/whisper_flutter_plus/assets/59436976/a83ec44f-8e7a-44cb-9690-7ca68fcd880b" width=400>

This is the speed comparison of running in Release mode, which is worse in debug mode.

**Comment by krikristoophe** [comment]
Hello,
From my tests, the transcription with my package is too slow to be usable in real time (in my own app, I display transcription to user only when available. Whisper work as a background task).

Also, the speed is very dependent on 2 things : the device used and the running mode (release is much much faster than debug). If you tried with the same configuration, I think that my package can me upgraded to be faster. I didn't know the demo app, if it's open source we can maybe find the difference

As reminder, this package is only an implementation of whisper.cpp, I'd just managed to make it work but didn't touch to whisper.cpp. I'm not really aware to the whisper functioning. If you find anything to make it work better or faster, feel free to make a PR or proposition !

Have a good day

**Comment by LinWhite2333** [comment]
I'm much faster with this demo 
https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.android
but I'm not sure what the difference is.

**Comment by shalehaha** [comment]
There might be some steps you need to do to get it optimized. I haven't tried the Android lib yet, but I did follow https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.objc to get the compiler optimized for my iOS project and it is now much faster. The translation time was down from ~30 secs to about only 2 secs.

**Comment by LinWhite2333** [comment]
Thank you for your answer, but there is no relevant information in the android library, I don't understand how to optimize in the android library

**Comment by jtkeyva** [comment]
cool @shalehaha do you have an example of this implemented with flutter?

**Comment by tempo-riz** [comment]
any update on this ? :)

**Comment by stopthinking102** [comment]
To get transcriptions in less than a second just set the app to release mode and 
 set this compiler flag ggml.c -DGGML_USE_ACCELERATE

**Comment by stopthinking102** [comment]
The compiler flag -DGGML_USE_ACCELERATE in ggml.c enables the use of Apple's Accelerate framework within the GGML (Georgi Gerganov's Machine Learning) library. This directive defines a preprocessor macro that instructs the compiler to leverage Accelerate's high-performance mathematical and image processing functions, which are optimized for Apple hardware. By using this flag, certain operations in GGML are implemented using Accelerate's BLAS and vDSP functions instead of their default implementations. This can result in significant performance improvements on Apple devices such as Macs, iPhones, and iPads, as the Accelerate framework is specifically optimized for Apple's processors. Essentially, this flag allows GGML to take advantage of hardware-specific optimizations provided by Apple, potentially enhancing the library's performance on Apple platforms.

**Comment by tempo-riz** [comment]
Hey @shalehaha @krikristoophe @stopthinking102 did you find a way to optimize the android version ? Otherwise we could consider using method channels to get android demo's perf ?

**Comment by stopthinking102** [comment]
> Hey @shalehaha @krikristoophe @stopthinking102 did you find a way to optimize the android version ? Otherwise we could consider using method channels to get android demo's perf ?

For what use case u want to optimize . long form audio or shortform /realtime audio on android

**Comment by stopthinking102** [comment]
Android is not gonna be supporting NNAPI 
Warning: With the release of Android 15, NNAPI will be deprecated. While you can continue to use NNAPI, we expect the majority of devices in the future to use the CPU backend, and therefore for performance critical workloads, we recommend migrating to alternative solutions, for example the TF Lite GPU runtime.

and whisper.cpp the inference engine this library uses is not optomized for say qualcomm chipset which most android phones use. There can be small gains by removing the overhead of serialization n deseralization n directly calling the function . Dont know why this library uses json to pass around audio data. U can load the model once n use it for subsequent requests. Also check the whisper threads parameter. Look up distill models they work much faster.

**Comment by tempo-riz** [comment]
> > Hey @shalehaha @krikristoophe @stopthinking102 did you find a way to optimize the android version ? Otherwise we could consider using method channels to get android demo's perf ?
> 
> For what use case u want to optimize . long form audio or shortform /realtime audio on android

To transcribe audio files of about 30 sec to 10min length mostly

So you say that whisper.cpp is not really for android ? Should I look into TF Lite ?

**Comment by stopthinking102** [comment]
whisper.cpp does not use NPUs and GPUs on android like it does on apple neural engine and support for metal on apple devices.
For now u can increase thread count , u can double the speed of audio like 60 second could be processed in the same window as 30 second audio with speed up of audio.
and build for android u can optimize with cpu features for android like

**https://github.com/ggerganov/whisper.cpp/blob/master/examples/whisper.android/lib/src/main/jni/whisper/CMakeLists.txt**
   if (${target_name} STREQUAL "whisper_v8fp16_va")
        target_compile_options(${target_name} PRIVATE -march=armv8.2-a+fp16)
        set(GGML_COMPILE_OPTIONS                      -march=armv8.2-a+fp16)
    elseif (${target_name} STREQUAL "whisper_vfpv4")
        target_compile_options(${target_name} PRIVATE -mfpu=neon-vfpv4)
        set(GGML_COMPILE_OPTIONS                      -mfpu=neon-vfpv4)
    endif ()

Going tflite route will speed up inference  but doesn't handle audio preprocessing (generating the mel spectrogram) or post-processing (translating logits to text).

Do share your findings

**Comment by jtkeyva** [comment]
> There might be some steps you need to do to get it optimized. I haven't tried the Android lib yet, but I did follow https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.objc to get the compiler optimized for my iOS project and it is now much faster. The translation time was down from ~30 secs to about only 2 secs.

How did you get this running in flutter?

**Comment by jtkeyva** [comment]
> To get transcriptions in less than a second just set the app to release mode and set this compiler flag ggml.c -DGGML_USE_ACCELERATE

hello, how and where do i set that for flutter? thx

**Comment by stopthinking102** [comment]
> > To get transcriptions in less than a second just set the app to release mode and set this compiler flag ggml.c -DGGML_USE_ACCELERATE
> 
> hello, how and where do i set that for flutter? thx

So ur flutter project will contain an IOS directory. Open the runner in xcode.
Then in xcode u can change the settings like shown in the screenshots on this page.
https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.objc

btw this only works on ios apple devices not on android.

**Comment by jtkeyva** [comment]
thank you!

**Comment by tempo-riz** [comment]
ok so i adapted whisper.cpp's android demo https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.android using method chanels on android, and here are my results for jfk.wav using base model

- this package (using ffi) : 174284 ms and  8354 ms in release mode
- using method chanels : 34451 ms and 2116 ms in release mode

So it's about 4 times faster in release 🤔

**Comment by stopthinking102** [comment]
using whisper.cpp with ffi that does it about in 2.3 seconds with the tiny model.
have u seen this tflite version https://github.com/nyadla-sys/whisper.tflite


Thing is that if u want to get any considerable speedup my guess is when u can utilize the gpu.
For this the model has to be supported by tflite. How to check for support here >>
https://ai.google.dev/edge/litert/models/ops_compatibility

May be what u can do is try with the repository shared by nyadia n see if u can compile n make it work with GPU if the model is compatible. And if u get any speed up.That would be more or less general .
Even further speed to use npu gpu u can look into qualcomm sdk
https://aihub.qualcomm.com/models/whisper_tiny_en

The results they have posted on their website makes it much close to the kind of performance u can get with metal and coreml on ios for android.

**Comment by tempo-riz** [comment]
I also tried this plugin : https://pub.dev/packages/whisper_flutter_new
it took 3586 ms in release mode, it's using ffi but with better compilation stpes/optimization I believe

**Comment by tempo-riz** [comment]
I tried nyada's version and sadly it's audio input are limited to 30s so we would need to preprocess the audio into chunks which would decrease transcript quality probalby...
I think it's related to tflite RAM limitations but I might be wrong

**Comment by tempo-riz** [comment]
> using whisper.cpp with ffi that does it about in 2.3 seconds with the tiny model. have u seen this tflite version https://github.com/nyadla-sys/whisper.tflite
> 
> Thing is that if u want to get any considerable speedup my guess is when u can utilize the gpu. For this the model has to be supported by tflite. How to check for support here >> https://ai.google.dev/edge/litert/models/ops_compatibility
> 
> May be what u can do is try with the repository shared by nyadia n see if u can compile n make it work with GPU if the model is compatible. And if u get any speed up.That would be more or less general . Even further speed to use npu gpu u can look into qualcomm sdk https://aihub.qualcomm.com/models/whisper_tiny_en
> 
> The results they have posted on their website makes it much close to the kind of performance u can get with metal and coreml on ios for android.


I tried nyada's version and sadly it's audio input are limited to 30s so we would need to preprocess the audio into chunks which would decrease transcript quality probalby...
I think it's related to tflite RAM limitations but I might be wrong

**Comment by tempo-riz** [comment]
but we don't have this limitation with either this plugin or whisper.cpp's android demo, do they split the audio ?

**Comment by tempo-riz** [comment]
are you planning to try to make gpu work ? Because it's really out of my expertise haha

**Comment by stopthinking102** [comment]
whisper splits the audio in whisper_full_parallel . The reason i am now thinking its not really useful to do this on the device is that android is just a mess. u have googe pixel chipsets, samsung exynos processors , qualcomm most widely used  n media thek and huwawei. And even after u get it to work u get to target high end devices.

For apple u get decent performance u can run a distilled large model with low word error rate. After u posted i got curious looked into it but for my usecase which is near to realtime audio transcription i feet its not worth it.

I might look into it when these are implemented
https://ai.google.dev/edge/litert/android/npu

Or may be we can ask nyada if he tried with gpu support

**Comment by stopthinking102** [comment]
for now thinking to use this https://groq.com/groq-runs-whisper-large-v3-at-a-164x-speed-factor-according-to-new-artificial-analysis-benchmark/

**Comment by tempo-riz** [comment]
thanks for the ideas, I heard about groq which sounds promising even tho not on device :)

**Comment by stopthinking102** [comment]
So turns out whisper does not have gpu / nnapi  support. 
Model uses dynamic tensors dat are not supported by tflite both gpu delegate and nnapi delegate fail.

**Comment by ltanedo** [comment]
> > > To get transcriptions in less than a second just set the app to release mode and set this compiler flag ggml.c -DGGML_USE_ACCELERATE
> > 
> > 
> > hello, how and where do i set that for flutter? thx
> 
> So ur flutter project will contain an IOS directory. Open the runner in xcode. Then in xcode u can change the settings like shown in the screenshots on this page. https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.objc
> 
> btw this only works on ios apple devices not on android.

Hey you guys did great work, but where is the correct path of the ggml.c file.  ggml.c isn't present in the compile sources table, did you guys have to add it, in order to set the -DGGML_USE_ACCELERATE flag?

![Screenshot 2025-01-08 at 1 48 37 AM](https://github.com/user-attachments/assets/207444a5-3820-4b1c-bdf3-8e4a3ea671f3)
